pkg/controller/controller.go

```diff
@@ -0,0 +1,242 @@
+		},
+	})
+
+	err := iamMappingInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+		"canonicalARN": IndexIAMIdentityMappingByCanonicalArn,
+	})
+	if err != nil {
+		logrus.WithError(err).Fatal("error adding index")
+	}
+
+	controller.iamMappingsIndex = iamMappingInformer.Informer().GetIndexer()
+
+	return controller
+}
+
+// Run will implement the loop for processing items
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	logrus.Info("starting aws iam authenticator controller")
+
+	logrus.Info("waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.iamMappingsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	logrus.Info("starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	logrus.Info("started workers")
+	<-stopCh
+	logrus.Info("shutting down workers")
+
+	return nil
+}
+
+// runWorker loops over each item looking for boolean returns
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will process each item off the queue
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+
+		if err := c.syncHandler(key); err != nil {
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing %s : %s, requeuing", key, err.Error())
+		}
+
+		c.workqueue.Forget(obj)
+		logrus.Infof("successfully synced %s", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (c *Controller) syncHandler(key string) (err error) {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key %s", key))
+		return nil
+	}
+
+	iamIdentityMapping, err := c.iamMappingLister.Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("iam identity mapping %s no longer exists", key))
+			return nil
+		}
+		return err
+	}
+
+	// Process items
+	if iamIdentityMapping.Spec.ARN != "" {
+		iamIdentityMappingCopy := iamIdentityMapping.DeepCopy()
+
+		canonicalizedARN, err := arn.Canonicalize(strings.ToLower(iamIdentityMapping.Spec.ARN))
+		if err != nil {
+			return err
+		}
+
+		iamIdentityMappingCopy.Status.CanonicalARN = canonicalizedARN
+		_, err = c.iamclientset.IamauthenticatorV1alpha1().IAMIdentityMappings().Update(iamIdentityMappingCopy)
+		if err != nil {
+			return err
+		}
+	}
+
+	c.recorder.Event(iamIdentityMapping, corev1.EventTypeNormal, SuccessSynced, IdentitySynced)
```

Comment:
Instead of catching an event here I made it so you can't have blank arns https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/116/commits/be527c14fc29794f4ea79103c10ea3a853f9c2d0#diff-29d9ef80094554ca8c330d50387f63dbR23